pkg/algorithm.py

```diff
@@ -0,0 +1,34 @@
+"""Module for Generic Algorithm class"""
+class Algorithm:
+    """Generic Algorithm class for algorithm factory
+    """
+    def __init__(self, matcher, dataframe, test):
+        self.matcher = matcher
+        self.dataframe = dataframe
+        self.test = test
+
+    def output_json(self):
```

Comment:
These method implementations needs to be enforced in the inherited classes, for this class to be a factory technically. We can achieve this with `@abstractmethod` decorator. And all the other methods can either be static or private ones thinking about asynchronous behavior.